BaragonData/src/main/java/com/hubspot/baragon/data/BaragonStateDatastore.java

```diff
@@ -143,50 +142,33 @@ public int getGlobalStateSize() {
   }
 
   private Map<String, Collection<UpstreamInfo>> fetchServiceToUpstreamInfoMap(Collection<String> services) throws Exception {
-    Map<String, String> upstreamToService = fetchUpstreamToServiceMap(services);
-
-    Collection<String> upstreamPaths = new ArrayList<>(upstreamToService.size());
-    for (Entry<String, String> entry : upstreamToService.entrySet()) {
-      String service = entry.getValue();
-      String servicePath = ZKPaths.makePath(SERVICES_FORMAT, service);
-
-      String upstream = entry.getKey();
-      String upstreamPath = ZKPaths.makePath(servicePath, upstream);
-
-      upstreamPaths.add(upstreamPath);
-    }
-
-    Map<String, UpstreamInfo> upstreamToInfo = zkFetcher.fetchDataInParallel(upstreamPaths, new UpstreamInfoDeserializer());
-
+    Map<String, Collection<String>> serviceToUpstreams = zkFetcher.fetchChildrenInParallel(services);
+    Map<String, UpstreamInfo> upstreamToInfo = zkFetcher.fetchDataInParallel(upstreamPaths(serviceToUpstreams), new UpstreamInfoDeserializer());
     Map<String, Collection<UpstreamInfo>> serviceToUpstreamInfo = new HashMap<>(services.size());
-    for (Entry<String, UpstreamInfo> entry : upstreamToInfo.entrySet()) {
-      String upstream = entry.getKey();
-      String service = upstreamToService.get(upstream);
-      Preconditions.checkNotNull(service, String.format("Invalid state for upstream '%s'", upstream));
-
-      if (serviceToUpstreamInfo.containsKey(service)) {
-        serviceToUpstreamInfo.get(service).add(entry.getValue());
-      } else {
-        serviceToUpstreamInfo.put(service, Lists.newArrayList(entry.getValue()));
+
+    for (Entry<String, Collection<String>> entry : serviceToUpstreams.entrySet()) {
+      for (String upstream : entry.getValue()) {
+        if (serviceToUpstreamInfo.containsKey(entry.getKey())) {
+          serviceToUpstreamInfo.get(entry.getKey()).add(upstreamToInfo.get(upstream));
+        } else {
+          serviceToUpstreamInfo.put(entry.getKey(), Lists.newArrayList(upstreamToInfo.get(upstream)));
+        }
       }
     }
 
     return serviceToUpstreamInfo;
   }
 
-  private Map<String, String> fetchUpstreamToServiceMap(Collection<String> services) throws Exception {
-    Map<String, Collection<String>> serviceToUpstreams = zkFetcher.fetchChildrenInParallel(services);
-
-    Map<String, String> upstreamToService = new HashMap<>();
+  private Collection<String> upstreamPaths(Map<String, Collection<String>> serviceToUpstreams) {
+    Collection<String> allUpstreamPaths = new ArrayList<>();
     for (Entry<String, Collection<String>> entry : serviceToUpstreams.entrySet()) {
-      String service = entry.getKey();
-
+      String servicePath = ZKPaths.makePath(SERVICES_FORMAT, entry.getKey());
       for (String upstream : entry.getValue()) {
-        upstreamToService.put(upstream, service);
+        String upstreamPath = ZKPaths.makePath(servicePath, upstream);
+        allUpstreamPaths.add(upstreamPath);
```

Comment:
would probably be simpler to just to `String.format(UPSTREAM_FORMAT, entry.getKey(), upstream)`